ethcore/engines/validator-set/src/safe_contract.rs

```diff
@@ -283,6 +317,55 @@ impl ValidatorSafeContract {
 			Some(matched_event) => Some(SimpleList::new(matched_event.new_set))
 		}
 	}
+
+	fn filter_report_queue(&self, our_address: &Address, latest_block: BlockNumber) -> Result<(), EthcoreError>{
+		let client = self.client.read().as_ref().and_then(Weak::upgrade).ok_or("No client!")?;
+		let client = client.as_full_client().ok_or("No full client!")?;
+
+		self.queued_reports.lock().unwrap().retain(|&(malicious_validator_address, block, ref _data)| {
+			trace!(target: "engine", "Checking if report can be removed from cache");
+			if block > latest_block {
+				return false; // Report cannot be used, as it is for a block that isn’t in the current chain
+			}
+			if latest_block > 100 && latest_block - 100 > block {
+				return false; // Report is too old and cannot be used
+			}
+			// Check if the validator is already banned...
```

Comment:
Just thinking out aloud here, but isn't it worthwhile to cut down on the number of contract calls? Could it make a single call like `does_this_guy_need_to_be_reported::call()? In other words: it doesn't seem like we care about the reason a report should be dropped.
Taking that one step further: would it be more performant to just send off all the reports we have and let the contract decide if the report is a duplicate or already processed?

Reply:
Sure, I'd be happy to replace them with a single call per report, or even with a single call for the whole list. I don't know how expensive constant local contract calls are, and what the overhead of several simple vs. a single complex call is.

@varasev: What do you think about this API change?

Reply:
OK, I added a separate getter to `ValidatorSet` contract, so let's replace [these lines](https://github.com/poanetwork/parity-ethereum/blob/dbdd9c5f0f0cbe274b7cc6962decc997e3ebe319/ethcore/engines/validator-set/src/safe_contract.rs#L592-L629) with a single call to [that getter](https://github.com/poanetwork/posdao-contracts/blob/2700b23fb4e6ad2a80802c72175c2a18b2149cb7/contracts/ValidatorSetAuRa.sol#L562-L586).

Reply:
Done.